.travis.yml

```diff
@@ -24,35 +12,69 @@ before_install:
   - wget https://github.com/swig/swig/archive/rel-3.0.12.tar.gz
   - tar xzf rel-3.0.12.tar.gz
   - cd swig-rel-3.0.12 && ./autogen.sh && ./configure --prefix=/usr && make && sudo make install && cd ..
-  - sudo pip install coverage
-  - sudo pip install coveralls
-  - echo "import coverage" | sudo tee --append /usr/lib/python2.7/sitecustomize.py
-  - echo "coverage.process_startup()" | sudo tee --append /usr/lib/python2.7/sitecustomize.py 
 
-script: 
-  - export TOTAL_BUILD_JOBS=8
-  - export COVERAGE_PROCESS_START=$PWD/.coveragerc
-  - echo "[run]" >> .coveragerc
-  - echo "source = $PWD/src" >> .coveragerc
-  - echo "parallel = True" >> .coveragerc
-# WORKAROUND: attempt to retry JobTests.py if it fails and then continue if it passes, if it fails ten times
-# then it is a real failure not related to intermittent travis failures
-  - n=0; while [[ $n -lt 10 ]]; do coverage run --rcfile=$PWD/.coveragerc runtest.py src/engine/SCons/JobTests.py && break; n=$((n+1)); done; if [ "$n" -gt "9" ]; then false; fi
-# exclude JobTest.py becuase we already ran that
-  - echo "src/engine/SCons/JobTests.py" > exclude_jobtest
-# also exclude this test since it overides the exit function which doesnt work with coverage 
-# more info here: https://coverage.readthedocs.io/en/coverage-4.4.2/subprocess.html#
-# TODO: figure out how to cover test/exitfns.py
-  - echo "test/exitfns.py" >> exclude_jobtest
-  - python runtest.py -l -a --exclude-list exclude_jobtest > all_tests
-  - let "start = ($(wc -l < all_tests) / ${TOTAL_BUILD_JOBS}) * (${BUILD_JOB_NUM} - 1)"; true;
-  - let "end = ($(wc -l < all_tests) / ${TOTAL_BUILD_JOBS}) * ${BUILD_JOB_NUM}"
-  - if (( ${BUILD_JOB_NUM} == ${TOTAL_BUILD_JOBS} )); then end=$(wc -l < all_tests); fi
-  - if (( ${start} == 0 )); then start=1; fi
-  - sed -n ${start},${end}p all_tests > build_tests
-  - coverage run --rcfile=$PWD/.coveragerc runtest.py -f build_tests || if [[ $? == 2 ]]; then true; else false; fi
+jobs:
+  include:
+    - &test_job
+      stage: Test
+      script: 
+        - n=0; while [[ $n -lt 10 ]]; do python runtest.py src/engine/SCons/JobTests.py && break; n=$((n+1)); done; if [ "$n" -gt "9" ]; then false; fi
+        - echo "src/engine/SCons/JobTests.py" > exclude_jobtest
+        - python runtest.py -a --exclude-list exclude_jobtest || if [[ $? == 2 ]]; then true; else false; fi
+      before_script: skip
+      after_success: skip
+
+    - &coverage_jobs
+      stage: Coverage
+      
+      before_script:
+        - sudo pip install coverage
+        - sudo pip install coveralls
+        - echo "import coverage" | sudo tee --append /usr/lib/python2.7/sitecustomize.py
```

Comment:
Can we updated runtest.py to import coverage if available? Would that be better?

Reply:
We would need to make some modifications to the SconsTest modules so they start their subprocesses with coverage enabled. This Travis script does it the quick and easy way but does have the drawback of needing to modify the python distribution.

For the time being I think this is good for getting coverage reports going for the project, while the only benefit to making the SconsTest modules use coverage directly would be the convenience of not needing to mess with the python distribution.

This is how coverage actually recommends you monitor subprocesses as well:
https://coverage.readthedocs.io/en/coverage-4.4.2/subprocess.html#
> Complex test suites may spawn sub-processes to run tests, either to run them in parallel, or because sub-process behavior is an important part of the system under test. Measuring coverage in those sub-processes can be tricky because you have to modify the code spawning the process to invoke coverage.py.
> 
> There’s an easier way to do it: coverage.py includes a function, coverage.process_startup() designed to be invoked when Python starts. It examines the COVERAGE_PROCESS_START environment variable, and if it is set, begins coverage measurement. The environment variable’s value will be used as the name of the configuration file to use.

Reply:
I'd be interested in being able to run coverage via runtest.py without having to run via travis/coveralls... But not needed to merge this logic. Thanks for the work!